Dockerfile

```diff
@@ -1,22 +1,89 @@
-FROM phusion/baseimage:0.11
-# developer Dockerfile for mesa development, installs from local git checkout
-LABEL maintainer="Allen Lee <allen.lee@asu.edu>"
+# The Mesa project publishes "official" code to:
+# - Pip: https://pypi.org/project/Mesa
+# - Docker Hub: Not yet!
 
-ENV PYTHONUNBUFFERED=1 \
-    LANG=C.UTF-8 \
-    LC_ALL=C.UTF-8
+# There is no official Mesa Docker image.  Use this example to build your own.
 
-WORKDIR /opt/mesa
+# You can build and run this Docker image with these commands:
+# $ docker build . -t mymesa_image
+# $ docker rm -f mymesa_instance
+# $ docker run --name mymesa_instance -p 8521:8521/tcp -it mymesa_image 
+# View at http://127.0.0.1:8521
 
-COPY . /opt/mesa
+# "bash" the instance like this:
+# $ docker exec -it mymesa_instance /bin/bash 
 
-RUN apt-get update && apt-get upgrade -y -o Dpkg::Options::="--force-confold" \
-    && apt-get install -y --no-install-recommends \
-    build-essential \
-    python3-dev \
-    python3-pip \
-    python3-setuptools \
-    python3-wheel \
-    && rm -rf /var/lib/apt/lists/*
+# FOR SECURITY:
+# - Don't use random Docker images, libraries, or napkins in the bin
+# - Use official Docker images, or images packaged by the same people that made the library.
+# - Otherwise, who knows what code is in that image?!
 
-RUN pip3 install -e /opt/mesa
+# This example Dockerfile packages the wolf_sheep example model.
+
+# To use your own model, place your code in /opt/mesa/mymodel. 
+# You have two options:
+
+# OPTION A) COPY files into Docker image
+
+# From your other codebase, do these steps:
+# 1) Copy this Dockerfile into your model codebase
+# 2) Replace `COPY examples/boid-flockers /opt/mesa/mymodel`
+#       with `COPY [dir w/ your run.py] /opt/mesa/mymodel`
+# 3) Follow steps to build and run above
+
+# OPTION B) Mount a volume using docker-compose or kubernetes
+
+# `docker-compose.yml` in this repo shows an example of this 
+# models mounted in volumes must have their dependencies managed outside docker
+
+
+# Example Dockerfile for wolf_sheep model
+# Steps in this file are ordered least likly to change (top) to most likely to change (bottom)
+
+# Use official python image from Docker team: https://hub.docker.com/_/python
+FROM python:stretch
+
+# Expose our port:
+EXPOSE 8521/tcp
+
+# Update OS
+RUN apt-get update && apt-get upgrade -y 
+
+# Don't buffer output: https://docs.python.org/3.7/using/cmdline.html?highlight=pythonunbuffered#envvar-PYTHONUNBUFFERED
+ENV PYTHONUNBUFFERED=1 
+# Cruft? Set locale to use UTF-8: http://www.iac.es/sieinvens/siepedia/pmwiki.php?n=Tutorials.LinuxLocale
+ENV LANG=C.UTF-8
+ENV LC_ALL=C.UTF-8
+
+# Install pipenv
+RUN pip install --upgrade pipenv
```

Comment:
`pipenv` is not yet used as the default installation method, is it?

Reply:
This file is an example.  Models use Mesa as a library.  So they may use pipenv for their dependencies.  This pipenv is for the person using the example, not Mesa.

Reply:
The file is a generic Dockerfile that can run any simulations that may have various install methods, be it in requirements.txt, setup.py, Pipfile, or pyproject.toml. In some cases where requirements.txt is not sufficient (e.g. `cythonize`), only `setup.py install` can be used. Since the `pipenv install` line is commented out, this (optional) line should, too.